README.md
# RFID reader  

Study project.  

Full RFID system with web interface.  

Main microcontroller	- STM32F100RB  
Web module		- esp8266  
RFID module		- SeeedStudio RDM 125KHz  
2.8' LCD display	- ILI9341 driver  
Touchpanel		- xpt2046 driver  
SD Card			- included in LCD display  
  
Features implemented:  
* displaing actual time, display turn off after time interval. Turn on by touch
* Sensing RFID cards  
* Specific respons to single RFID card displayed on LCD. Response bases on saved permission list
* Making and saving history of sensed cards (with date of this event)  
* Read only access to history from web  
* Read and write access to permission list from web  


sd_data/history.txt
21.05.2016,18:17:01,1003AC0DA2;21.05.2016,18:20:21,1003AC0DA2;21.05.2016,18:27:41,1003AC0DA2;21.05.2016,18:27:48,1003AC0DA2;21.05.2016,18:28:19,1003AC0DA2;21.05.2016,18:29:57,10030AA55C;22.05.2016,14:08:02,1003AC0DA2;22.05.2016,14:08:11,1003AC0DA2;22.05.2016,14:08:21,1003AC0DA2;30.05.2016,20:27:56,1003AC0DA2;30.05.2016,20:28:07,10030AA55C;30.05.2016,20:28:20,10030AA55C;30.05.2016,20:28:27,1003AC0DA2;30.05.2016,20:28:33,1003AC0DA2;30.05.2016,20:29:07,1003AC0DA2;30.05.2016,20:29:14,10030AA55C;01.06.2016,07:20:37,1003AC0DA2;01.06.2016,07:21:58,10030AA55C;01.06.2016,07:29:50,10030AA55C;01.06.2016,07:30:41,1003AC0DA2;02.06.2016,18:23:20,1003AC0DA2;14.06.2016,13:48:48,1003AC0DA2;14.06.2016,13:48:56,10030AA55C;14.06.2016,13:50:15,1003AC0DA2;14.06.2016,13:50:22,1003AC0DA2;14.06.2016,13:50:31,10030AA55C;14.06.2016,13:50:37,10030AA55C;14.06.2016,13:50:43,10030AA55C;14.06.2016,13:50:48,10030AA55C;14.06.2016,13:51:58,10030AA55C;14.06.2016,13:54:04,1003AC0DA2;14.06.2016,13:54:09,1003AC0DA2;14.06.2016,14:28:22,10030AA55C;14.06.2016,14:28:28,1003AC0DA2;14.06.2016,16:10:24,1003AC0DA2;14.06.2016,16:49:58,10030AA55C;14.06.2016,16:55:30,10030AA55C;14.06.2016,19:43:55,10030AA55C;14.06.2016,19:48:16,10030AA55C;14.06.2016,19:48:23,1003AC0DA2;14.06.2016,20:20:09,1003AC0DA2;14.06.2016,20:20:35,1003AC0DA2;14.06.2016,20:20:50,1003AC0DA2;14.06.2016,20:20:56,10030AA55C;14.06.2016,20:21:56,10030AA55C;14.06.2016,20:22:24,1003AC0DA2;14.06.2016,20:22:30,1003AC0DA2;14.06.2016,20:22:51,10030AA55C;14.06.2016,20:23:12,10030AA55C;14.06.2016,20:26:52,10030AA55C;14.06.2016,20:28:12,1003AC0DA2;14.06.2016,21:57:28,1003AC0DA2;14.06.2016,21:57:36,10030AA55C;14.06.2016,21:57:46,10030AA55C;14.06.2016,22:08:44,10030AA55C;14.06.2016,22:08:51,1003AC0DA2;14.06.2016,22:15:44,10030AA55C;14.06.2016,22:15:56,10030AA55C;14.06.2016,22:16:02,10030AA55C;14.06.2016,22:16:09,10030AA55C;14.06.2016,22:31:37,10030AA55C;14.06.2016,22:31:44,1003AC0DA2;14.06.2016,22:32:00,1003AC0DA2;14.06.2016,22:32:06,10030AA55C;15.06.2016,09:08:09,1003AC0DA2;15.06.2016,09:09:36,10030AA55C;15.06.2016,09:09:43,1003AC0DA2;15.06.2016,09:13:16,10030AA55C;15.06.2016,09:34:02,10030AA55C;15.06.2016,09:34:09,1003AC0DA2;15.06.2016,09:34:16,10030AA55C;15.06.2016,09:34:28,1003AC0DA2;15.06.2016,09:34:33,10030AA55C;15.06.2016,09:35:01,10030AA55C;15.06.2016,09:35:17,1003AC0DA2;15.06.2016,09:35:26,1003AC0DA2;15.06.2016,09:35:34,1003AC0DA2;15.06.2016,09:35:47,1003AC0DA2;15.06.2016,10:24:09,10030AA55C;15.06.2016,10:24:15,1003AC0DA2;15.06.2016,10:26:03,10030AA55C;15.06.2016,10:26:09,1003AC0DA2;15.06.2016,10:26:43,10030AA55C;15.06.2016,10:26:49,1003AC0DA2;15.06.2016,11:37:18,10030AA55C;15.06.2016,11:37:24,1003AC0DA2;15.06.2016,11:48:08,10030AA55C;15.06.2016,11:48:45,1003AC0DA2;15.06.2016,11:48:57,1003AC0DA2;17.06.2016,18:06:02,10030AA55C;17.06.2016,18:06:08,10030AA55C;17.06.2016,18:06:17,1003AC0DA2;17.06.2016,18:06:25,10030AA55C;17.06.2016,18:06:37,10030AA55C;17.06.2016,18:06:55,1003AC0DA2;17.06.2016,18:07:01,1003AC0DA2;17.06.2016,18:07:08,10030AA55C;17.06.2016,18:07:13,10030AA55C;17.06.2016,18:13:49,1003AC0DA2;17.06.2016,18:13:55,10030AA55C;17.06.2016,18:14:02,1003AC0DA2;17.06.2016,19:15:38,10030AA55C;17.06.2016,19:15:45,1003AC0DA2;17.06.2016,19:15:58,1003AC0DA2;17.06.2016,19:44:41,1003AC0DA2;17.06.2016,19:44:47,10030AA55C;17.06.2016,20:28:11,10030AA55C;17.06.2016,20:28:21,1003AC0DA2;17.06.2016,20:29:04,10030AA55C;17.06.2016,20:41:05,1003AC0DA2;17.06.2016,20:41:11,10030AA55C;20.04.2130,14:52:32,10030AA55C;20.04.2130,14:52:39,1003AC0DA2;20.04.2130,14:53:24,1003AC0DA2;20.04.2130,14:53:30,10030AA55C;20.04.2130,14:53:56,1003AC0DA2;20.04.2130,14:54:03,10030AA55C;20.04.2130,14:54:08,10030AA55C;20.04.2130,14:54:13,1003AC0DA2;20.04.2130,14:54:27,1003AC0DA2;19.02.2016,10:36:13,10030AA55C;19.02.2016,10:36:23,1003AC0DA2;19.02.2016,09:59:31,10030AA55C;19.02.2016,11:39:20,10030AA55C;19.02.2016,11:39:27,1003AC0DA2;19.02.2016,11:51:29,10030AA55C;19.02.2016,11:51:36,1003AC0DA2;19.02.2016,11:51:42,10030AA55C;19.02.2016,11:52:34,10030AA55C;19.02.2016,11:52:43,1003AC0DA2;19.02.2016,11:52:55,10030AA55C;19.02.2016,11:53:01,1003AC0DA2;19.02.2016,11:53:15,1003AC0DA2;19.02.2016,11:53:20,1003AC0DA2;19.02.2016,11:53:26,1003AC0DA2;19.02.2016,11:53:41,10030AA55C;19.02.2016,11:53:54,10030AA55C;19.02.2016,11:54:01,1003AC0DA2;19.02.2016,11:54:08,1003AC0DA2;19.02.2016,11:54:25,1003AC0DA2;19.02.2016,11:54:39,10030AA55C;19.02.2016,11:54:50,1003AC0DA2;19.02.2016,11:56:50,10030AA55C;19.02.2016,11:56:57,1003AC0DA2;19.02.2016,11:57:10,1003AC0DA2;19.02.2016,11:57:16,10030AA55C;19.02.2016,11:57:24,1003AC0DA2;19.02.2016,11:57:30,10030AA55C;19.02.2016,11:58:48,1003AC0DA2;19.02.2016,12:00:29,1003AC0DA2;19.02.2016,12:01:14,1003AC0DA2;19.02.2016,12:02:46,10030AA55C;19.02.2016,12:02:54,1003AC0DA2;19.02.2016,12:03:01,1003AC0DA2;19.02.2016,12:03:08,1003AC0DA2;19.02.2016,12:03:13,1003AC0DA2;19.02.2016,12:03:19,1003AC0DA2;19.02.2016,12:03:25,1003AC0DA2;19.02.2016,12:03:30,1003AC0DA2;19.02.2016,12:03:36,1003AC0DA2;19.02.2016,12:03:42,1003AC0DA2;19.02.2016,12:03:47,1003AC0DA2;19.02.2016,12:03:53,1003AC0DA2;19.02.2016,12:04:41,777FC8E8;19.02.2016,12:13:54,1003AC0DA2;19.02.2016,12:14:00,1003AC0DA2;19.02.2016,12:14:04,1003AC0DA2;19.02.2016,12:14:10,1003AC0DA2;19.02.2016,12:14:16,1003AC0DA2;19.02.2016,12:14:21,1003AC0DA2;19.02.2016,12:14:27,1003AC0DA2;19.02.2016,12:14:33,1003AC0DA2;19.02.2016,12:14:38,1003AC0DA2;19.02.2016,12:14:45,1003AC0DA2;19.02.2016,12:14:49,1003AC0DA2;19.02.2016,12:14:56,1003AC0DA2;19.02.2016,12:15:01,1003AC0DA2;19.02.2016,12:15:06,1003AC0DA2;19.02.2016,12:15:12,1003AC0DA2;19.02.2016,12:15:18,1003AC0DA2;19.02.2016,12:15:23,1003AC0DA2;19.02.2016,12:18:21,1003AC0DA2;19.02.2016,12:18:27,1003AC0DA2;19.02.2016,12:18:33,1003AC0DA2;19.02.2016,12:18:38,1003AC0DA2;19.02.2016,12:18:44,1003AC0DA2;19.02.2016,12:18:50,1003AC0DA2;19.02.2016,12:18:55,1003AC0DA2;19.02.2016,12:19:38,1003AC0DA2;19.02.2016,12:19:44,1003AC0DA2;19.02.2016,12:19:49,1003AC0DA2;19.02.2016,12:19:55,1003AC0DA2;19.02.2016,12:20:00,1003AC0DA2;19.02.2016,12:20:

RFID_src/RTC.c
#include "RTC.h"

__IO uint8_t RTC_second_flag = 0;
extern __IO uint8_t esp8266_second_flag;

void RTC_IRQHandler() {
	static uint8_t test = 0;
	test++;
	RTC_second_flag = 1;
	if (test == 40) {
		esp8266_second_flag = 1;
		test = 0;
	}

	RTC->CRL &= ~RTC_CRL_SECF;
}

static uint8_t RTC_YearIsLeap(uint16_t year)
{
	if (((year / 4) * 4 == year) && (year % 400 == 0))
		return 1;
	else if (((year / 4) * 4 == year) && (year % 100 != 0))
		return 1;

	return 0;
}

static uint8_t RTC_DaysInMonth(uint16_t year, uint8_t month)
{
	if (month == 1 || month == 3 || month == 5 || month == 7 || month == 8
			 || month == 10 || month == 12) 
		return 31;
	else if (RTC_YearIsLeap(year) && month == 2)
		return 29;
	else if (month == 2)
		return 28;
	else
		return 30;

	return 0;
}

void RTC_SetDate(uint16_t year, uint8_t month, uint8_t day, uint8_t hour, 
		 uint8_t min, uint8_t sec)
{
	uint32_t days = 0;
	uint8_t months = 1;
	uint32_t years = 0;
	uint32_t temp = 0;
	uint32_t cnt = 2016;

	years = year - cnt;

	for (uint32_t i = 0; i < years; i++) {
		if (RTC_YearIsLeap(cnt))
			days += 366;
		else
			days += 365;
		
		cnt++;
	}

	days += day;

	if (hour >= 24) {
		days++;
		hour -= 24;
	}

	while (months < month) {
		days += RTC_DaysInMonth(cnt, months);
		months++;
	}

	temp = days * 60 * 60 * 24 + hour * 60 * 60 + min * 60 + sec;
	/* Entering configuration mode */
	RTC->CRL |= RTC_CRL_CNF;
	/* Waiting for entry to configuration mode */
	while(!(RTC->CRL & RTC_CRL_RTOFF));

	RTC->CNTH = (uint16_t)(temp >> 16);
	RTC->CNTL = (uint16_t)(temp);
	/* Leaving configuration mode */
	RTC->CRL &= ~RTC_CRL_CNF;
	
	/* Waiting for end of write operation to CRL*/
	while(!(RTC->CRL & RTC_CRL_RTOFF)); 
}

void RTC_SetTime(uint8_t hour, uint8_t min, uint8_t sec) {
	uint32_t temp = hour * 3600 + min * 60 + sec;
	/* Entering configuration mode */
	RTC->CRL |= RTC_CRL_CNF;
	/* Waiting for entry to configuration mode */
	while(!(RTC->CRL & RTC_CRL_RTOFF));

	RTC->CNTH = (uint16_t)(temp >> 16);
	RTC->CNTL = (uint16_t)(temp);
	/* Leaving configuration mode */
	RTC->CRL &= ~RTC_CRL_CNF;
	
	/* Waiting for end of write operation to CRL*/
	while(!(RTC->CRL & RTC_CRL_RTOFF)); 
}

uint8_t RTC_GetTime(uint8_t *hour, uint8_t *min, uint8_t *sec) {
	uint32_t temp;
	uint8_t ret = 0;
	static uint8_t hour_s, min_s, sec_s;
	if (RTC_second_flag) {
		temp = (RTC->CNTH << 16) | RTC->CNTL;
		hour_s = (temp / 3600) % 24;
		min_s = temp / 60 % 60;
		sec_s = temp % 60;
		ret = 1;
		RTC_second_flag = 0;
	}
	*hour = hour_s;
	*min = min_s;
	*sec = sec_s;
	return ret;
}

static void RTC_CalculateDate(uint32_t temp, uint8_t *month_s, uint8_t *day_s, 
			 uint8_t *hour_s, uint8_t *min_s, uint8_t *sec_s, 
			 uint16_t *year_s)
{
	uint8_t not_leap = 0;
	*year_s = 0;
	*month_s = 0;
	*day_s = 0;
	*sec_s = temp % 60;
	temp -= *sec_s;

	*min_s = (temp / 60) % 60;
	temp -= *min_s * 60;

	*hour_s = (temp / 3600) % 24;
	temp -= *hour_s * 3600;
		
	temp = temp / (60 * 60 * 24);

	*year_s = 2016;
	while(temp) {
		if (*month_s == 0) {
			*month_s = 1;
		}
		else if (*month_s == 1 || *month_s == 3 || *month_s == 5 || *month_s == 7 || *month_s == 8
			 || *month_s == 10 || *month_s == 12) {
			if (temp > 31) {
				*month_s = *month_s + 1;
				if (*month_s == 13) {
					*month_s = 1;
					not_leap++;
					not_leap %= 4;
					year_s++;
				}
				temp -= 31;	
			} else {
				*day_s = temp;
				temp = 0;
			}
		}

		else if (*month_s == 2) {
			if (not_leap && (temp > 28)) {
			*month_s = *month_s + 1;
				temp -= 28;
			} else if (not_leap) {
				*day_s = temp;
				temp = 0;	
			} else if (!not_leap && (temp > 29)) {
				*month_s = *month_s + 1;
				temp -= 29;
			} else {
				*day_s = temp;
				temp = 0;
			}
		}

		else if (*month_s == 4 || *month_s == 6 || *month_s == 9 || *month_s == 11) {
			if (temp > 30) {
				*month_s = *month_s + 1;
				temp -= 30;
			} else {
				*day_s = temp;
				temp = 0;
			}
		}	
	}
}

static uint8_t RTC_GetLastSunday(uint16_t year, uint8_t month, uint8_t day)
{
	uint16_t tmp, tmp_2 = 0;
	const uint8_t months_val[] = {0, 0, 3, 3, 6, 1, 4, 6, 2,5, 0, 3, 5};
	uint8_t wod;
	tmp = day + months_val[month];
	tmp %= 7;
	tmp_2 = (year % 100) % 28;
	tmp_2 += (year % 100) / 4;

	if (RTC_YearIsLeap(year) && (month == 1 || month == 2))
		tmp_2--;

	wod = (tmp + tmp_2) % 7;

	if (wod)
		return 0;

	if (day + 7 > RTC_DaysInMonth(year, month)) {
		return day; 
	} else {
		while (day + 7 < RTC_DaysInMonth(year, month)) 
			day += 7;

		return day;
	}

	return 0;
}

static uint8_t RTC_CheckDST(uint32_t *temp, uint16_t year, uint8_t month, 
			    uint8_t day, uint8_t hour)
{
	uint8_t tmp;

	if (month > 10 || month < 3) {
		*temp += 60 * 60;
	} else if (month < 10 && month > 3) {
		*temp += 2 * 60 * 60;
	} else if (month == 3) {
		tmp = RTC_GetLastSunday(year, month, day);
		if (day < tmp) {
			*temp += 60 * 60;
		} else if (day > tmp) {
			*temp += 2 * 60 * 60;
		} else {
			if (hour < 1)
				*temp += 60 * 60;
			else
				*temp += 2 * 60 * 60;
		}
	} else {
		tmp = RTC_GetLastSunday(year, month, day);
		if (day < tmp) {
			*temp += 2 * 60 * 60;
		} else if (day > tmp) {
			*temp += 60 * 60;
		} else {
			if (hour < 1)
				*temp += 2 * 60 * 60;
			else
				*temp += 60 * 60;
		}
	}
	return 0;	
}

uint8_t RTC_GetDate(uint16_t *year, uint8_t *month, uint8_t *day, 
		    uint8_t *hour, uint8_t *min, uint8_t *sec) {
	uint32_t temp;
	uint8_t ret = 0;
	static uint8_t month_s, day_s, hour_s, min_s, sec_s;
	static uint16_t year_s;
	if (RTC_second_flag) {

		temp = (RTC->CNTH << 16) | RTC->CNTL;

		RTC_CalculateDate(temp, &month_s, &day_s, &hour_s, &min_s, &sec_s, &year_s);
		RTC_CheckDST(&temp, year_s, month_s, day_s, hour_s);
		RTC_CalculateDate(temp, &month_s, &day_s, &hour_s, &min_s, &sec_s, &year_s);		

		ret = 1;
		RTC_second_flag = 0;
	}
	*year = year_s;
	*month = month_s;
	*day = day_s;
	*hour = hour_s;
	*min = min_s;
	*sec = sec_s;
	return ret;
}

HAL_StatusTypeDef RTC_Init()
{
	/*
	 * After reset access to RTC registers is disabled, so
	 * there is need to update RCC register. Look to datasheet
	 */
	RCC->APB1ENR |= RCC_APB1ENR_PWREN | RCC_APB1ENR_BKPEN;
	PWR->CR |= PWR_CR_DBP;
	// Setting clock to RTC. Using LSE clock (low speed external)
	RCC->BDCR |= RCC_BDCR_RTCSEL_0;
	RCC->BDCR &= RCC_BDCR_RTCSEL_1;
	// Enabling LSE oscillator
	RCC->BDCR |= RCC_BDCR_LSEON;
	// Waiting for stable oscillator
	while(!(RCC->BDCR & RCC_BDCR_LSERDY));
	// Enabling RTC clock;
	RCC->BDCR |= RCC_BDCR_RTCEN;	
	/* Synchronizing APB1 bus with RTC (read datasheet)*/
	while(!(RTC->CRL & RTC_CRL_RSF));
	/* Waiting for end of write operation */
	while(!(RTC->CRL & RTC_CRL_RTOFF));
	/* Entering configuration mode */
	RTC->CRL |= RTC_CRL_CNF;
	/* Second interrupt enable */
	RTC->CRH = RTC_CRH_SECIE;
	/* RTC prescaler reload value high */
	RTC->PRLH = 0;
	/* RTC prescaler reload value low */
	RTC->PRLL = 0x7FFF;
	/* Leaving configuration mode */
	RTC->CRL &= ~RTC_CRL_CNF;
	/* Waiting for end of write operation */
	while(!(RTC->CRL & RTC_CRL_RTOFF)); 
	HAL_NVIC_EnableIRQ(RTC_IRQn);
	return HAL_OK;
}
